File: patterns/01-design-patterns/03-behavioral/06-observer/python/observer.py
class PatternError(Exception):
    def __init__(self,code):super().__init__(code);self.code=code
class Observer:
    def __init__(self,name):self.name=name
    def notify(self,event):
        if self.name=="failing":raise PatternError("observer_failed")
class Subject:
    def __init__(self,observers):self.observers=list(observers)
    def publish(self,event):
        receipts=[]
        for observer in list(self.observers):
            try:observer.notify(event);receipts.append({"event":event,"observer":observer.name,"status":"delivered"})
            except PatternError as error:receipts.append({"event":event,"observer":observer.name,"status":"failed","error":error.code})
        return receipts
    def unsubscribe(self,names):self.observers=[value for value in self.observers if value.name not in names]
def evaluate(input_data):
    supported={"audit","metrics","failing"}
    if any(name not in supported for name in input_data.get("observers",[])):raise PatternError("unsupported_observer")
    subject=Subject([Observer(name) for name in input_data.get("observers",[])]);receipts=[]
    for index,event in enumerate(input_data.get("events",[])):
        receipts.extend(subject.publish(event))
        if index==0:subject.unsubscribe(set(input_data.get("unsubscribe_after_first",[])))
    return{"receipts":receipts,"active":len(subject.observers)}

File: patterns/01-design-patterns/03-behavioral/06-observer/go/observer.go
package observer

type PatternError struct{ code string }

func (e *PatternError) Error() string { return e.code }
func (e *PatternError) Code() string  { return e.code }

type Observer interface {
	Name() string
	Notify(string) error
}
type namedObserver struct{ name string }

func (o namedObserver) Name() string { return o.name }
func (o namedObserver) Notify(string) error {
	if o.name == "failing" {
		return &PatternError{code: "observer_failed"}
	}
	return nil
}

type Subject struct{ observers []Observer }

func (s *Subject) Publish(event string) []map[string]any {
	out := make([]map[string]any, 0, len(s.observers))
	for _, o := range append([]Observer(nil), s.observers...) {
		if err := o.Notify(event); err != nil {
			out = append(out, map[string]any{"event": event, "observer": o.Name(), "status": "failed", "error": err.(*PatternError).Code()})
		} else {
			out = append(out, map[string]any{"event": event, "observer": o.Name(), "status": "delivered"})
		}
	}
	return out
}
func (s *Subject) Unsubscribe(names map[string]bool) {
	kept := []Observer{}
	for _, o := range s.observers {
		if !names[o.Name()] {
			kept = append(kept, o)
		}
	}
	s.observers = kept
}
func Evaluate(input map[string]any) (any, error) {
	raw, _ := input["observers"].([]any)
	subject := &Subject{}
	for _, name := range raw {
		value := name.(string)
		if value != "audit" && value != "metrics" && value != "failing" {
			return nil, &PatternError{code: "unsupported_observer"}
		}
		subject.observers = append(subject.observers, namedObserver{value})
	}
	receipts := []map[string]any{}
	events, _ := input["events"].([]any)
	for index, event := range events {
		receipts = append(receipts, subject.Publish(event.(string))...)
		if index == 0 {
			names := map[string]bool{}
			if values, ok := input["unsubscribe_after_first"].([]any); ok {
				for _, v := range values {
					names[v.(string)] = true
				}
			}
			subject.Unsubscribe(names)
		}
	}
	return map[string]any{"receipts": receipts, "active": len(subject.observers)}, nil
}
